articles/containerization-and-why-to-use-it.md

```diff
@@ -0,0 +1,58 @@
+---
+authorName: Florian Besser
+authorGithubUsername: florian-besser
+issue: 76
+title: Containerization and why to use it
+---
+# {{page.title}}
+
+You are a good software engineer, but you dread releases as they are known as a time of pain. Your project matches one or more of the following:
+* Deployments are done manually or via scripts that execute on some remote host.
+* Scaling the app is hard as new machines or VMs take weeks to arrive & setup.
+* Your application dependencies differ in version or feature set between environments.
+* Developers have difficulty reproducing production issues on their local machines.
+* Application dependencies are installed manually, developers spend days setting up their own machine.
+
+This article will explain how containerization can help solve each of these issues.
+
+## Cumbersome deployments and dirty hosts
+
+Running scripts on a host seems OK in the beginning, but over time your host will become "dirty". Some versions of your script were only run in QA, and production has leftover artifacts from previous deployments. These leftovers confuse maintainers and reduce confidence in your team's ability to deploy software to production systems.
+
+With containerization, you immediately get rid of leftovers your software may cause. By **throwing away the old container** after deploying a new version you have already cleaned up after your application.
+
+A tangent about volumes/host mounts: Sometimes your app writes files outside of its container. In these cases the files still need to be cleaned up, but usually an app has a well-defined output directory that rarely changes.
+
+Volumes allow you to use containerization with databases quite nicely: The data is safe from crashes but up- and downgrading a DB is just a question of stopping the old container and booting a new one.
+
+## More hardware and the procurement process
+
+There is a conflict between needing more resources and a company's willingness to spend money. Usually, there is a long-winded procurement process for ordering hardware and getting it installed it in a rack. In today's time where hardware is cheap and humans are expensive paying some PaaS a few hundred bucks a month is cheaper in most cases.
+
+Any established platform will offer you support for Docker, and if you have containerized your application you'll be able to solve the scaling problem a lot easier. Usually, this involves uploading the image to the platform, creating a service and then defining scaling rules for said service. While the platform will not run on your local machine this does not prevent you from testing locally, as you can install Docker and are guaranteed the same runtime as on the platform.
```

Comment:
Das habe ich ich jetzt in dem generellen Docker Hype noch nicht verstanden: Warum ist scaling easier wenn ich Docker verwende als wenn ich plain auf eine Cloud Infrastruktur setze. Nach meinem Verständnis ist es komplizierter.

cloud + docker: cloud server müssen raufskalieren und darauf müssen die docker images deplyoed werden
cloud: wir ein zusätzlicher server benötigt, wird der hinzugefügt.

Ich denke nicht, dass diese Diskussion/Unterscheidung noch Platz im Artikel findet. Aber es wäre mal spannend da drüber zu diskutieren.

Reply:
Und wie setzt du denn im traditionellen Modell den neuen Server auf und garantierst, dass dort die gleiche Applikation und der gleiche Build läuft? Schlussendlich muss da ja auch ein Deployment step dabei sein und ein produktiven Server willst du typischerweise auch nicht einfach so klonen